ext/pcntl/pcntl.c

```diff
@@ -1476,6 +1481,115 @@ PHP_FUNCTION(pcntl_setns)
 }
 #endif
 
+#ifdef HAVE_SCHED_SETAFFINITY
+PHP_FUNCTION(pcntl_getcpuaffinity)
+{
+	zend_long pid;
+	bool pid_is_null = 1;
+	cpu_set_t mask;
+
+	ZEND_PARSE_PARAMETERS_START(0, 1)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_LONG_OR_NULL(pid, pid_is_null)
+	ZEND_PARSE_PARAMETERS_END();
+
+	// 0 == getpid in this context, we re just saving a syscall
+	pid = pid_is_null ? 0 : pid;
+
+	CPU_ZERO(&mask);
+
+	if (sched_getaffinity(pid, sizeof(mask), &mask) != 0) {
+		PCNTL_G(last_error) = errno;
+		switch (errno) {
+			case ESRCH:
+				zend_argument_value_error(1, "invalid process (" ZEND_LONG_FMT ")", pid);
+				RETURN_THROWS();
+			case EPERM:
+				php_error_docref(NULL, E_WARNING, "Calling process not having the proper privileges");
+				break;
+			default:
+				php_error_docref(NULL, E_WARNING, "Error %d", errno);
+		}
+
+		RETURN_EMPTY_ARRAY();
+	}
+
+	zend_ulong maxcpus = (zend_ulong)sysconf(_SC_NPROCESSORS_ONLN);
+	array_init(return_value);
+
+	for (zend_ulong i = 0; i < maxcpus; i ++) {
+		if (CPU_ISSET(i, &mask)) {
+			add_next_index_long(return_value, i);
+		}
+	}
+}
+
+PHP_FUNCTION(pcntl_setcpuaffinity)
+{
+	zend_long pid;
+	bool pid_is_null = 1;
+	cpu_set_t mask;
+	zval *hmask = NULL, *ncpu;
+
+	ZEND_PARSE_PARAMETERS_START(0, 2)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_LONG_OR_NULL(pid, pid_is_null)
+		Z_PARAM_ARRAY(hmask)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (!hmask || zend_hash_num_elements(Z_ARRVAL_P(hmask)) == 0) {
+		zend_argument_value_error(2, "must not be empty");
+		RETURN_THROWS();
+	}
+
+	// 0 == getpid in this context, we re just saving a syscall
+	pid = pid_is_null ? 0 : pid;
+	zend_ulong maxcpus = (zend_ulong)sysconf(_SC_NPROCESSORS_ONLN);
+	CPU_ZERO(&mask);
+
+	ZEND_HASH_FOREACH_VAL(Z_ARRVAL_P(hmask), ncpu) {
+		ZVAL_DEREF(ncpu);
+		zval cpu;
+		ZVAL_COPY_VALUE(&cpu, ncpu);
+		if (Z_TYPE(cpu) != IS_LONG) {
+			if (Z_TYPE(cpu) == IS_STRING) {
+				convert_to_long(&cpu);
```

Comment:
This is wrong. Note that the old value stored in the zval is destroyed when it gets converted. That means for non-interned strings, the reference count of that string will decrement by one, causing a use-after-free later on.
You should use `zval_get_long` on the string, without taking a copy probably.